echopype/convert/parse_azfp.py

```diff
@@ -127,25 +90,42 @@ def load_AZFP_xml(self):
         """Parses the AZFP  XML file.
         """
 
-        def get_value_by_tag_name(tag_name, element=0):
-            """Returns the value in an XML tag given the tag name and the number of occurrences."""
-            return px.getElementsByTagName(tag_name)[element].childNodes[0].data
-
         xmlmap = fsspec.get_mapper(self.xml_path, **self.storage_options)
-        px = xml.dom.minidom.parse(xmlmap.fs.open(xmlmap.root))
-
-        # Retrieve integer parameters from the xml file
-        for old_name, new_name in XML_INT_PARAMS.items():
-            self.parameters[new_name] = int(get_value_by_tag_name(old_name))
-        # Retrieve floating point parameters from the xml file
-        for param in XML_FLOAT_PARAMS:
-            self.parameters[param] = float(get_value_by_tag_name(param))
-        # Retrieve frequency dependent parameters from the xml file
-        for old_name, new_name in XML_FREQ_PARAMS.items():
-            self.parameters[new_name] = [
-                float(get_value_by_tag_name(old_name, ch))
-                for ch in range(self.parameters["num_freq"])
-            ]
+        root = ET.parse(xmlmap.fs.open(xmlmap.root)).getroot()
+
+        for child in root.iter():
+            camel_case_tag = self._to_camelCase(child.tag)
+            if len(child.attrib) > 0:
+                for key, val in child.attrib.items():
+                    self.parameters[camel_case_tag + "_" + self._to_camelCase(key)].append(val)
+
+            if all(char == "\n" for char in child.text):
+                continue
+            else:
+                try:
+                    val = int(child.text)
+                except ValueError:
+                    val = float(child.text)
+
+            if len(child.tag) > 3 and not child.tag.startswith("VTX"):
+                self.parameters[camel_case_tag].append(val)
+            else:
+                self.parameters[child.tag].append(val)
+
+        # Handling the case where there is only one value for each parameter
+        for key, val in self.parameters.items():
+            if len(val) == 1:
+                self.parameters[key] = val[0]
+
+    def _to_camelCase(self, tag):
```

Comment:
Actually, you're converting *from* CamelCase *to* snake_case; so, the function name should be changed.

More generally, conversion of strings between CamelCase and snake_case is a really common need, in Python and all languages. I was curious about existing implementations in Python (I thought it might even be in a standard library, but it wasn't). But -- unsurprisingly :stuck_out_tongue_winking_eye: -- it's already done elsewhere in echopype, for the EK parser:
https://github.com/OSOceanAcoustics/echopype/blob/3ca028f04a4fd3643bb0ad350dc09a699883ee1e/echopype/convert/utils/ek_raw_parsers.py#L709

There, it's wrapped up within a class.

Later, in a follow-up PR, @leewujung I think we should create a generic, stand-alone utility function to do this, then reuse it both here and with EK's. For this PR, I think it'd extend its scope too much. In the meantime, I think the EK implementation is more comprehensive. I tested that function and the one implemented here. For cases like "firstLast" (first character is not uppercase) and "FirstSecond_last" (mixed form), your function drops the parts that are not strictly CamelCase and returns "last" and "first_second", respectively. The `from_CamelCase` EK function returns "first_last" and "first_second_last", as expected. I found an [even more compact implementation](https://rodic.fr/blog/camelcase-and-snake_case-strings-conversion-with-python/) that also behaves like `from_CamelCase`. Your implementation also throws away numbers, or worse.

My recommendation for this PR: copy the `from_CamelCase` function `convert/utils/ek_raw_parsers.py` exactly as is, here. Note that there's no need for it to have the `self` argument, since it's not used.

Having identical functions will make it easier for the next PR that will move the functions to a utility function.

Reply:
Looks like I submitted my comment just as you were renaming and moving the function!

Reply:
@emiliom : I was about to merge this. Could you do what you said above, but instead of copying that function, move that function from `ek_raw_parsers.py` to `echopype/utils/misc.py`? We'll add other things to that module later, for example from #1124. After you add that I'll merge this PR. Thanks.

Reply:
Will do. I'm working on it.